### PR TITLE
tree: clean up ProcessPlaceholderAnnotations

### PR DESCRIPTION
**Pull request by RaduBerinde**
I found this code very hard to follow, especially the two
annotationState flags. This change makes some improvements:
 - the two flags are replaced with a more explicit state;
 - the error is generated where it is detected;
 - the types map is updated directly.

Release note: None

### CODE REVIEW

**Review by nvb — APPROVED** [review]
:lgtm: this is a nice simplification.

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33650#-:-LVuqpqn9dTHIuRH6Cwa:b3sdn4y)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/type_check.go, line 2118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33650#-LVupCfx6rf5o5Uu85aY:-LVupCfy3s4m9fvYhneq:b-sf62v7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7f400db8b4f73a2666a6542cb670b5545338551/pkg/sql/sem/tree/type_check.go#L2118)):*
> ```Go
> 						arg.Idx, v.types[arg.Idx],
> 					))
> 				}
> ```

Might as well have a `default` case to assert that you've exhaustively covered the states.

The other switch statements here don't try to be exhaustive. I usually find that empty branches and a default that panics provide a bit more safety, but it's up to you.
___
*[pkg/sql/sem/tree/type_check.go, line 2142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33650#-LVuqJLQ22vVZvODK49c:-LVuqJLQ22vVZvODK49d:bkhe3yx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7f400db8b4f73a2666a6542cb670b5545338551/pkg/sql/sem/tree/type_check.go#L2142)):*
> ```Go
> 	case *Placeholder:
> 		switch v.state[t.Idx] {
> 		case noType, typeFromCast:
> ```

Why does `noType` bring us to the `conflictingCasts` state instead of the `typeFromCast` state?

**Comment by RaduBerinde** [comment]
___
*[pkg/sql/sem/tree/type_check.go, line 2142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33650#-LVuqJLQ22vVZvODK49c:-LVuy64W0Ipz4-0sO7Tj:b-mq0vno) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7f400db8b4f73a2666a6542cb670b5545338551/pkg/sql/sem/tree/type_check.go#L2142)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why does `noType` bring us to the `conflictingCasts` state instead of the `typeFromCast` state?
</blockquote></details>

The cast thing is meant to only work if all placeholder occurrences are under (the same) cast. This is a placeholder that's not inside a cast. The `conflictingCasts` desciption mentions this: `either we found> conflicting casts, or we found an appearance of the placeholder without a cast`. I'll add a comment - this switch was definitely the most confusing in the existing code.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33650)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/type_check.go, line 2118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33650#-LVupCfx6rf5o5Uu85aY:-LVx7JW2AKvzZDZRY6es:bl60eq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7f400db8b4f73a2666a6542cb670b5545338551/pkg/sql/sem/tree/type_check.go#L2118)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Might as well have a `default` case to assert that you've exhaustively covered the states.

The other switch statements here don't try to be exhaustive. I usually find that empty branches and a default that panics provide a bit more safety, but it's up to you.
</blockquote></details>

Done.  The empty branches were good opportunity for some more comments.
